### DISCUSSION TITLE
messense/rust-musl-cross:armv7-musleabihf issue

### DISCUSSION THREAD

**Discussion by TheChiefMeat** [dani-garcia/vaultwarden#1621] [Q&A] [answered]
Hi all, I'm having a slight issue building the 1.20.0 server from source and was wondering if anyone knew why this is happening. I'm using the armv7 alpine image which I've been using for over a year now.

I'm getting this error:

```
 => CANCELED [internal] load metadata for docker.io/balenalib/armv7hf-alpine:3.13                                                                                                                               1.0s
 => CANCELED [internal] load metadata for docker.io/bitwardenrs/web-vault@sha256:8747cfaa2c6d87d1749e119dd884697e8099389aa9aca30a4d73d4ff796fe0e4                                                               1.0s
 => ERROR [internal] load metadata for docker.io/messense/rust-musl-cross:armv7-musleabihf                                                                                                                      1.0s
------
 > [internal] load metadata for docker.io/messense/rust-musl-cross:armv7-musleabihf:
------
error: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: no match for platform in manifest sha256:45360d95fd1e5dfc884365aff0855395650525e2de9a02a1cf4a485cccf01805: not found

```

I checked the latest release of the messense/rust-musl-cross:armv7-musleabihf and it appears that the image isn't being built for armv7 specifically, only amd64 and arm64:

![image](https://user-images.githubusercontent.com/17432768/115461149-01a2d080-a221-11eb-994b-260b63b544aa.png)

Is this why this issue is happening or could it be something else? 

The Bitwarden server is running on a RaspberryPi 4.

**Comment by BlackDex** [comment]
The rust-musl-cross images are only build for amd64 as far as i know. That they now support arm64 is new to me, and looking at the build-process it doesn't look he builds the toolchain specifically for arm64.

The rust-musl-cross images are only used to cross-build the binaries from an amd64 host to arm targets.
So, building this on a armv7 cpu will not work using these images.

Also, isn't a Rpi4 an aarch64 CPU?

**Reply by TheChiefMeat** [reply]
Well I've been building Bitwarden_RS with the armv7 apline dockerfile for months now and it was working fine up to version 1.19.0, I still have the image on my local docker server and running on the Pi itself.

Also the rust-musl-cross image specifies in the title that it's supposed to be for armv7 but hasn't been compiled for it; It might be an error on their side compiling it for arm64 instead of armv7.

Edit: Just checked the armv7 dockerfile and I can't see any changes that would indicate that this is on Bitwarden's end, so I have to assume that the rust-musl-cross image isn't being compiled properly for armv7 anymore.

**Comment by BlackDex** [comment]
The title indicates it can be used to cross build for armv7 on an amd64 host. Not that that image is too be used on an armv7 cpu.

Only thing i can think of is that you have had some qemu binfmt running on the arm host?

The change they did was to support arm64 arch as a host also instead of only amd64.

**Reply by TheChiefMeat** [reply]
I am building on an AMD64 host. I'm using docker buildx to build the image like so:

`docker buildx build --platform linux/arm/v7 -t registry.raspberrypi.local/bitwarden:arm -f docker/armv7/Dockerfile.alpine . --push`

This method was working but it just doesn't want to work anymore.

**Comment by BlackDex** [comment, marked as answer]
Checkout the build scripts in the repo. https://github.com/dani-garcia/bitwarden_rs/tree/master/hooks

Maybe they could help.

**Reply by TheChiefMeat** [reply]
It appears you can now just run the following command and it will work for armv7:

```
docker build -t registry.raspberrypi.local/bitwarden-armv7 -f docker/armv7/Dockerfile.alpine .
```

Thanks for the help and pointing me towards the hooks folder :)
